File: examples_py/examples_py/generators/jerk_limited_generator.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Jerk-limited trajectory adapter (point-to-point API).

Python mirror of
``examples_cpp/src/generators/jerk_limited_generator.cpp``.

At each transition, ``on_waypoint_changed()`` runs the offline S-curve
simulator with ``[current_pos, next_waypoint]``; ``evaluate()`` then
samples the resulting trajectory using a segment-local time
``t - t_start``.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import math
import os
from dataclasses import dataclass
from typing import Optional

import numpy as np
import yaml
from mavpy.model import State
from trajectory_generator_jerk_limited import (
    EndWaypoint,
    TrajectoryGenerator,
    Waypoint,
    _NativeGeneratorConfig,
)

from examples_py.framework import (
    ExampleConfig,
    ITrajectoryGenerator,
    ReferenceField,
    ReferenceSample,
    make_mask,
)


_MIN_HORIZONTAL_SPEED_FOR_YAW = 0.5   # [m/s]
_MAX_YAW_RATE_REF_RAD_PER_SEC = 1.0   # [rad/s]


@dataclass
class JerkLimitedConfig:
    max_acceleration: float = 0.0
    max_jerk: float = 0.0
    #: Forwarded to TrajectoryParameters but unused by the one-shot API
    #: (the streaming time-stretch gate does not apply to offline planning).
    #: Kept for schema compatibility; <= 0 disables the bound.
    max_tracking_error: float = 0.0


def _quat_to_yaw(q: np.ndarray) -> float:
    w, x, y, z = float(q[0]), float(q[1]), float(q[2]), float(q[3])
    siny_cosp = 2.0 * (w * z + x * y)
    cosy_cosp = 1.0 - 2.0 * (y * y + z * z)
    return math.atan2(siny_cosp, cosy_cosp)


def _wrap_to_pi(x: float) -> float:
    while x > math.pi:
        x -= 2.0 * math.pi
    while x < -math.pi:
        x += 2.0 * math.pi
    return x


class JerkLimitedGenerator(ITrajectoryGenerator):
    """Offline S-curve replanned per waypoint transition (p2p)."""

    def __init__(self, cfg: JerkLimitedConfig) -> None:
        self._cfg = cfg
        self._ctrl: Optional[TrajectoryGenerator] = None
        self._target_wp = np.zeros(3, dtype=float)
        self._hold_pos = np.zeros(3, dtype=float)
        self._duration = 0.0
        self._t_segment_start = 0.0
        self._has_plan = False
        self._max_speed = 0.0
        self._path_facing = False
        self._has_prev_t = False
        self._prev_t = 0.0
        self._yaw_ref_hold = 0.0
        self._name = 'JerkLimitedGenerator'

    @staticmethod
    def load_config_from_yaml(path: str) -> JerkLimitedConfig:
        if not os.path.isfile(path):
            raise ValueError(f'Config file not found at {os.path.abspath(path)}.')
        with open(path, 'r') as f:
            root = yaml.safe_load(f)
        if root is not None and not isinstance(root, dict):
            raise ValueError('jerk_limited config: root must be a mapping.')
        cfg = JerkLimitedConfig()
        if isinstance(root, dict):
            for key in ('max_acceleration', 'max_jerk', 'max_tracking_error'):
                if key in root:
                    setattr(cfg, key, float(root[key]))
        return cfg

    def initialize(self, initial_state: State, example_cfg: ExampleConfig) -> None:
        if example_cfg.max_speed <= 0.0:
            raise ValueError(
                'JerkLimitedGenerator: ExampleConfig.max_speed must be > 0 '
                '(set in config_example.yaml).')
        self._path_facing = bool(example_cfg.path_facing)
        self._has_prev_t = False
        self._prev_t = 0.0
        self._yaw_ref_hold = _quat_to_yaw(
            np.asarray(initial_state.orientation, dtype=float))
        self._max_speed = float(example_cfg.max_speed)

        p0 = np.asarray(initial_state.position, dtype=float)
        self._hold_pos = p0.copy()
        self._target_wp = p0.copy()
        self._duration = 0.0
        self._t_segment_start = 0.0
        self._has_plan = False

        # Construct the offline solver once; reused across segments via generate().
        cfg = _NativeGeneratorConfig()
        cfg.params.max_speed = self._max_speed
        cfg.params.max_acceleration = self._cfg.max_acceleration
        cfg.params.max_jerk = self._cfg.max_jerk
        cfg.params.max_tracking_error = self._cfg.max_tracking_error
        self._ctrl = TrajectoryGenerator(cfg)

    def on_waypoint_changed(
        self, next_waypoint: np.ndarray, state: State, t_start: float,
    ) -> None:
        if self._ctrl is None:
            raise RuntimeError(
                'JerkLimitedGenerator: initialize() must be called before '
                'on_waypoint_changed().')
        p0 = np.asarray(state.position, dtype=float)
        next_wp = np.asarray(next_waypoint, dtype=float).copy()
        self._target_wp = next_wp
        self._hold_pos = next_wp.copy()
        self._t_segment_start = float(t_start)

        if float(np.linalg.norm(next_wp - p0)) < 1e-6:
            self._has_plan = False
            self._duration = 0.0
            return

        # Two-waypoint hop. Pin the C1 initial conditions of the segment
        # to the live drone state (v0 = state.linear_velocity, a0 = 0) so
        # the jerk-limited integrator stitches continuously across replans
        # — matching the gcopter adapter's `wp[0].velocity = ...` pattern
        # and the aerostack2 plugin behaviour. Without this, every replan
        # reactive to a follow_reference modify injects a step in the
        # commanded velocity.
        start = Waypoint(p0)
        start.velocity = np.asarray(state.linear_velocity, dtype=float).copy()
        wps = [start, EndWaypoint(next_wp)]
        self._has_plan = bool(self._ctrl.generate(wps, self._max_speed))
        if not self._has_plan:
            self._duration = 0.0
            return
        self._duration = float(self._ctrl.duration())

    def update(self, t: float, state: State) -> None:
        del state
        if self._ctrl is None:
            raise RuntimeError(
                'JerkLimitedGenerator: initialize() must be called before update().')

        dt = max(t - self._prev_t, 0.0) if self._has_prev_t else 0.0
        self._prev_t = t
        self._has_prev_t = True

        if not self._has_plan:
            return
        t_local = max(0.0, min(t - self._t_segment_start, self._duration))

        if self._path_facing:
            vel = np.asarray(self._ctrl.velocity(t_local), dtype=float)
            horiz_speed = float(np.linalg.norm(vel[:2]))
            if horiz_speed > _MIN_HORIZONTAL_SPEED_FOR_YAW:
                yaw_target = math.atan2(float(vel[1]), float(vel[0]))
                yaw_delta = _wrap_to_pi(yaw_target - self._yaw_ref_hold)
                step = _MAX_YAW_RATE_REF_RAD_PER_SEC * dt
                yaw_delta = max(-step, min(step, yaw_delta))
                self._yaw_ref_hold += yaw_delta
        else:
            self._yaw_ref_hold = 0.0

    def evaluate(self, t: float) -> ReferenceSample:
        sample = ReferenceSample(yaw=self._yaw_ref_hold)
        if not self._has_plan or self._ctrl is None:
            sample.position = self._hold_pos.copy()
            return sample
        t_rel = t - self._t_segment_start
        t_eval = max(0.0, min(t_rel, self._duration))
        sample.position = np.asarray(self._ctrl.position(t_eval), dtype=float).copy()
        sample.velocity = np.asarray(self._ctrl.velocity(t_eval), dtype=float).copy()
        sample.acceleration = np.asarray(
            self._ctrl.acceleration(t_eval), dtype=float).copy()

        # Past the segment end, freeze on the exact target with zero motion.
        # The integrator stops once ‖v‖ ≤ settle_velocity, which can leave
        # the last grabbed sample a few centimetres short of the waypoint;
        # pin the outgoing reference to the requested target so the
        # controller does not see a residual offset during the hover plateau.
        if t_rel >= self._duration:
            sample.position = self._target_wp.copy()
            sample.velocity = np.zeros(3, dtype=float)
            sample.acceleration = np.zeros(3, dtype=float)
        return sample

    def provided_reference_fields(self) -> ReferenceField:
        return make_mask([
            ReferenceField.POSITION,
            ReferenceField.VELOCITY,
            ReferenceField.ACCELERATION,
        ])

    def name(self) -> str:
        return self._name

File: examples_cpp/src/generators/jerk_limited_generator.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file jerk_limited_generator.cpp
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include "generators/jerk_limited_generator.hpp"

#include <yaml-cpp/yaml.h>

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <utility>
#include <vector>

#include "utils/example_config_utils.hpp"
#include "utils/utils.hpp"

namespace mpc_examples::adapters {

namespace tg = trajectory_generator_jerk_limited;

namespace {

constexpr double kMinHorizontalSpeedForYaw = 0.5;  // [m/s]
constexpr double kMaxYawRateRefRadPerSec   = 1.0;  // [rad/s]

double wrapToPi(double x) {
  while (x > M_PI) x -= 2.0 * M_PI;
  while (x < -M_PI) x += 2.0 * M_PI;
  return x;
}

}  // namespace

JerkLimitedGenerator::JerkLimitedGenerator(const Config& cfg) : cfg_(cfg) {}

JerkLimitedGenerator::Config JerkLimitedGenerator::loadConfigFromYaml(const std::string& path) {
  const YAML::Node root = detail::loadYamlRoot(path);
  Config cfg;
  const auto read_opt = [&](const char* key, double& dst) {
    if (root[key]) {
      dst = detail::readDoubleRequired(root[key], key);
    }
  };
  read_opt("max_acceleration", cfg.max_acceleration);
  read_opt("max_jerk", cfg.max_jerk);
  read_opt("max_tracking_error", cfg.max_tracking_error);
  return cfg;
}

void JerkLimitedGenerator::initialize(const mav_model::State& initial_state,
                                      const ExampleConfig& example_cfg) {
  if (example_cfg.max_speed <= 0.0) {
    throw std::invalid_argument(
        "JerkLimitedGenerator: ExampleConfig::max_speed must be > 0 (set in config_example.yaml).");
  }
  path_facing_  = example_cfg.path_facing;
  has_prev_t_   = false;
  prev_t_       = 0.0;
  yaw_ref_hold_ = quaternionToEuler(initial_state.getOrientationVector()).z();
  max_speed_    = example_cfg.max_speed;

  hold_pos_        = initial_state.getPositionVector();
  target_wp_       = hold_pos_;
  duration_        = 0.0;
  t_segment_start_ = 0.0;
  has_plan_        = false;

  // Construct the offline solver once; it is reused across segments via generate().
  tg::GeneratorConfig generator_cfg;
  generator_cfg.params.max_speed          = example_cfg.max_speed;
  generator_cfg.params.max_acceleration   = cfg_.max_acceleration;
  generator_cfg.params.max_jerk           = cfg_.max_jerk;
  generator_cfg.params.max_tracking_error = cfg_.max_tracking_error;
  ctrl_ = std::make_unique<tg::TrajectoryGenerator>(generator_cfg);
}

void JerkLimitedGenerator::onWaypointChanged(const Eigen::Vector3d& next_waypoint,
                                             const mav_model::State& state,
                                             double t_start) {
  const Eigen::Vector3d p0 = state.getPositionVector();
  target_wp_               = next_waypoint;
  hold_pos_                = next_waypoint;
  t_segment_start_         = t_start;

  if ((next_waypoint - p0).norm() < 1e-6) {
    has_plan_ = false;
    duration_ = 0.0;
    return;
  }

  // Two-waypoint hop. The one-shot API resets internally to waypoints[0]
  // with v=0, a=0 — same as the rest of p2p adapters (gcopter, dynamic,
  // mav_traj_gen). The settle margin in the WaypointScheduler absorbs the
  // resulting v0 discontinuity at segment boundaries.
  std::vector<tg::Waypoint> wps = {tg::Waypoint(p0), tg::EndWaypoint(next_waypoint)};
  has_plan_                     = ctrl_->generate(wps, max_speed_);
  if (!has_plan_) {
    duration_ = 0.0;
    return;
  }
  duration_ = ctrl_->duration();
}

void JerkLimitedGenerator::update(double t, const mav_model::State& /*state*/) {
  const double dt = has_prev_t_ ? std::max(t - prev_t_, 0.0) : 0.0;
  prev_t_         = t;
  has_prev_t_     = true;

  if (!has_plan_) {
    return;
  }
  const double t_local = std::clamp(t - t_segment_start_, 0.0, duration_);

  if (path_facing_) {
    const Eigen::Vector3d vel = ctrl_->velocity(t_local);
    const double horiz_speed  = vel.head<2>().norm();
    if (horiz_speed > kMinHorizontalSpeedForYaw) {
      const double yaw_target = std::atan2(vel.y(), vel.x());
      double yaw_delta        = wrapToPi(yaw_target - yaw_ref_hold_);
      const double step       = kMaxYawRateRefRadPerSec * dt;
      yaw_delta               = std::clamp(yaw_delta, -step, step);
      yaw_ref_hold_           = yaw_ref_hold_ + yaw_delta;
    }
  } else {
    yaw_ref_hold_ = 0.0;
  }
}

framework::ReferenceSample JerkLimitedGenerator::evaluate(double t) const {
  framework::ReferenceSample sample;
  sample.yaw = yaw_ref_hold_;

  if (!has_plan_) {
    sample.position = hold_pos_;
    return sample;
  }
  const double t_rel  = t - t_segment_start_;
  const double t_eval = std::clamp(t_rel, 0.0, duration_);
  sample.position     = ctrl_->position(t_eval);
  sample.velocity     = ctrl_->velocity(t_eval);
  sample.acceleration = ctrl_->acceleration(t_eval);

  // Past the segment end, freeze on the exact target with zero motion.
  // The integrator stops once ‖v‖ ≤ settle_velocity, which can leave the
  // last grabbed sample a few centimetres short of the waypoint; pin the
  // outgoing reference to the requested target so the controller does not
  // see a residual offset during the hover plateau.
  if (t_rel >= duration_) {
    sample.position = target_wp_;
    sample.velocity.setZero();
    sample.acceleration.setZero();
  }
  return sample;
}

}  // namespace mpc_examples::adapters
